test: miscellaneous additions to coverage tests

galtm opened the pull request:
This pull request adds various test cases to the suite of end-to-end tests for XSLT code coverage:

- Non-XSLT children of `xsl:stylesheet` and `xsl:transform`
- Variations on the child nodes of `xsl:param` for template parameters
- Cases that include `as` attribute in `xsl:param` and `xsl:variable`
- Various ways of creating text nodes
- Parents not hit, for certain elements where we apply or might start to apply the "use parent data" rule (`xsl:context-item`, `xsl:with-param`, `xsl:sort`)

When creating names for new items in existing test files, I sometimes used words instead of numbers because keeping all the numbers in order would make the code and its history churn more than necessary. I'm open to feedback if you think any of the choices here make the tests confusing.

Cc: @birdya22

birdya22 commented on test/end-to-end/cases-coverage/xsl-param-02.xsl:
Can you change the template match, or something similar, because is it the same match as for xsl-param-01.xsl and trying to include both in the same test causes problems with duplicate template matches.

galtm replied:
Good catches, thanks. I made updates in 49ac664ddf0e03f0763ca113203c3b26e6d5403e.

birdya22 commented on test/end-to-end/cases-coverage/xsl-param-02.xspec:
See comment on xsl-param-02.xsl

birdya22 commented on test/end-to-end/cases-coverage/expected/stylesheet/xsl-param-01-coverage.html:
Stylesheet numbering lines 10, 32 and 36 in are shown as 'ignored' but their content is shown as 'missed'. The comments say 'Expected miss' so I would expect the 3 `xsl:param ...` to be shown as 'missed'.
There are no hits in the trace file for those lines.

birdya22 commented:
Before I add any more comments, can I check whether the html coverage report should reflect any 'Expected miss' comments  because I'm seeing some elements marked as 'ignored' when I would expect them to be 'missed'.

galtm commented:
> Before I add any more comments, can I check whether the html coverage report should reflect any 'Expected miss' comments because I'm seeing some elements marked as 'ignored' when I would expect them to be 'missed'.

Yes. In addition to addressing your feedback so far, I'll wait to get a green light from you before I merge this pull request.

birdya22 commented on test/end-to-end/cases-coverage/expected/stylesheet/xsl-param-01-coverage.html:
Stylesheet numbering lines 10, 32 and 36 are marked as 'ignored' but their content is shown as 'missed'. The comments say 'Expected miss' so I would expect the 3 `xsl:param ...` entries to be shown as 'missed'.
There are no hits in the trace file for those lines.

galtm replied:
I think the "ignored" marking is something that should be fixed in `coverage-report.xsl`, coming from the line that says it's about top-level non-XSLT elements but actually casts a wider net.

https://github.com/xspec/xspec/blob/67644fcd7b2c4ff3489afc2ec9d0f26f89a15e98/src/reporter/coverage-report.xsl#L371-L372

birdya22 commented on test/end-to-end/cases-coverage/expected/stylesheet/xsl-variable-01-coverage.html:
Stylesheet lines 18-30 inclusive are shown as 'hit' but the comments say 'Expected miss' and there is no hit in the trace file for any of these lines.

Stylesheet lines 44-57 inclusive demonstrate the difficulty with `xsl:variable` in a `xsl:template`. All the `xsl:variable` elements have a 'hit' element in the trace file even though they are not used. 
Stylesheet lines 47, 48 and 48 are marked as 'hit' but there is no 'hit' element in the trace file.
The best thing is probably to wait and see what Saxon 12.5 with the fix for [6415](https://saxonica.plan.io/issues/6415) delivers before doing any more of this.

galtm replied:
> Stylesheet lines 18-30 inclusive are shown as 'hit' but the comments say 'Expected miss' and there is no hit in the trace file for any of these lines.

I think the word "Expected" in the comment is a bit aspirational and doesn't mean "this is what the code does, and the code is correct." I believe the "hit" markings are coming from

https://github.com/xspec/xspec/blob/67644fcd7b2c4ff3489afc2ec9d0f26f89a15e98/src/reporter/coverage-report.xsl#L363-L368

Saxon 12.5 was just released. With that version, `xsl-variable-01-coverage.xml` does not have 'hit' elements in the trace file for the unused `xsl:variable` elements in stylesheet lines 44-57 inclusive. Maybe that, along with changes in `coverage-report.xsl`, will enable us to mark the unused template variables as "missed".

I also noticed that Saxon 12.5 doesn't have a hit in the trace for the `xsl:variable` elements that define `variableLocalEmptySequence01` and `variableLocalEmptyString01` (lines 41-42) even though these variables _are_ used.

birdya22 replied:
Yes you are right about 'Expected'. I should have asked a better question to start with.
I've had a quick look at 2 reports side by side of most of the coverage tests and my main observation is that `xsl:copy` no longer seems to be traced, plus, as you noticed there are some changes for `xsl:variable`. I didn't see any new trace data.
How do you want to manage the changes related to Saxon 12.5?

birdya22 replied:
I've had a look at issue-214 and the demo ones and don't have anything to add. 
I've been writing some notes on the differences I can see in the trace output and will share when I've finished it.

birdya22 replied:
On a slight tangent I've discovered that it is possible to use the Tlevel settings to get different amounts of trace out of your own XSLTTraceListener as long as it extends AbstractTraceListener (I had to look at the Saxon code to work that out). 
I've basically got a copy of the standard Saxon XSLTTraceLsitener at the moment.

Generally there seems to be very little difference between the trace for 'normal' and 'high' in terms of the lines/columns traced but I did try a `xsl:copy` element with a select attribute. and that gets traced with 'high', but not with 'normal'.

I need to make a XSpec TraceListener to see what that does.

birdya22 replied:
FYI the original `xsl-merge-01.xsl` test fails with a Saxon error when running it with Tlevel:high. See Saxonica issue [6467](https://saxonica.plan.io/issues/6467).

galtm replied:
> Generally there seems to be very little difference between the trace for 'normal' and 'high' in terms of the lines/columns traced but I did try a `xsl:copy` element with a select attribute. and that gets traced with 'high', but not with 'normal'.
> 
> I need to make a XSpec TraceListener to see what that does.

Interesting. Just a bit of background that may or may not be relevant -- I don't have sufficient Java skills to do much with the XSpec trace listener. If you find that modifying the trace listener code is helpful, that's great, but my participation in discussion/review of the Java code will be pretty limited. (Other contributors to this repo have Java skills.)

birdya22 commented:
I've completed my comments.

galtm commented:
b059fb4ac14a65d5fe4d6f7863ee00774b46e6bc adds an unhit `xsl:merge` element to the test for that element, in support of #1959. The diff is a little noisy because I also made the indentation consistent in the file `xsl-merge-01.xsl`.

galtm commented:
@birdya22 , thanks for creating a separate issue (#1963) for discussion of Saxon 12.5 - good move. Do you think this pull request needs to wait for something specific before I merge it?

birdya22 commented:
No comments on this:

> https://github.com/xspec/xspec/commit/b059fb4ac14a65d5fe4d6f7863ee00774b46e6bc adds an unhit xsl:merge element to the test for that element, in support of https://github.com/xspec/xspec/issues/1959. The diff is a little noisy because I also made the indentation consistent in the file xsl-merge-01.xsl.

No on this:

> @birdya22 , thanks for creating a separate issue (https://github.com/xspec/xspec/issues/1963) for discussion of Saxon 12.5 - good move. Do you think this pull request needs to wait for something specific before I merge it?
